File: lang/java/src/main/java/ru/dz/mqtt_udp/util/LoopRunner.java
package ru.dz.mqtt_udp.util;

import java.io.IOException;

import ru.dz.mqtt_udp.MqttProtocolException;

/**
 * Set up thread, run user's code in a loop. Provide start and stop controls.
 * 
 * @author dz
 *
 */
public abstract class LoopRunner {

	// ------------------------------------------------------------
	// Constructor
	// ------------------------------------------------------------

	/**
	 * 
	 * @param name Thread name
	 */
	public LoopRunner(String name) {
		threadName = name;
	}
	
	// ------------------------------------------------------------
	// For user to override
	// ------------------------------------------------------------

	/**
	 * To be overridden in subclass. On start do preparations needed.
	 * 
	 * @throws IOException In case of IO error
	 * 
	 * @throws MqttProtocolException In case of MQTT/UDP protocol error
	 */
	protected abstract void onStart() throws IOException, MqttProtocolException;

	/**
	 * To be overridden in subclass. Called in loop to do actual work.
	 * 
	 * @throws IOException In case of IO error
	 * 
	 * @throws MqttProtocolException In case of MQTT/UDP protocol error
	 */
	protected abstract void step() throws IOException, MqttProtocolException;

	/**
	 * To be overridden in subclass. On stop do cleanup needed.
	 * 
	 * @throws IOException In case of IO error
	 * 
	 * @throws MqttProtocolException In case of MQTT/UDP protocol error
	 */
	protected abstract void onStop() throws IOException, MqttProtocolException;




	// ------------------------------------------------------------
	// Incoming data process thread
	// ------------------------------------------------------------

	volatile private boolean run = false;
	private final String threadName;

	public boolean isRunning() { return run; }


	/**
	 * Request to start reception loop thread.
	 */
	public void requestStart()
	{
		if(isRunning()) return;
		start();
	}

	/**
	 * Request to stop reception loop thread.
	 */
	public void requestStop() { run = false; }

	/**
	 * Worker: start loop thread.
	 */
	protected void start() {
		Runnable target = makeLoopRunnable();
		Thread t = new Thread( target, threadName );
		t.start();
	}


	private void loop() throws IOException, MqttProtocolException 
	{
		
		onStart();

		run = true;

		while(run)
		{
			step();
		}

		onStop();
	}


	private Runnable makeLoopRunnable() {
		return new Runnable() {
			@Override
			public void run() {
				try {
					loop();
				} catch (IOException e) {
					GlobalErrorHandler.handleError(ErrorType.IO, e);
				} catch (MqttProtocolException e) {
					GlobalErrorHandler.handleError(ErrorType.Protocol, e);
				}				
			}
		};
	}

	
	public static void sleep(long msec)
	{
		try {
			Thread.sleep(msec);
		} catch (InterruptedException e) {
			// Ignore
		}		
	}

}

File: lang/kotlin/mqtt_udp/src/main/kotlin/dz/ru/mqtt_udp/kt/util/LoopRunner.kt
package dz.ru.mqtt_udp.kt.util

import dz.ru.mqtt_udp.kt.GlobalErrorHandler
import java.io.IOException
import kotlin.concurrent.Volatile


/**
 * Set up thread, run user's code in a loop. Provide start and stop controls.
 *
 * @param threadName Thread name
 *
 * @author dz
 *
 */

abstract class LoopRunner(val threadName : String) : Sleeper() {


    // ------------------------------------------------------------
    // For user to override
    // ------------------------------------------------------------
    /**
     * To be overridden in subclass. On start do preparations needed.
     *
     * @throws IOException In case of IO error
     *
     * @throws MqttProtocolException In case of MQTT/UDP protocol error
     */
    @Throws(IOException::class, MqttProtocolException::class)
    protected abstract fun onStart()

    /**
     * To be overridden in subclass. Called in loop to do actual work.
     *
     * @throws IOException In case of IO error
     *
     * @throws MqttProtocolException In case of MQTT/UDP protocol error
     */
    @Throws(IOException::class, MqttProtocolException::class)
    protected abstract fun step()

    /**
     * To be overridden in subclass. On stop do cleanup needed.
     *
     * @throws IOException In case of IO error
     *
     * @throws MqttProtocolException In case of MQTT/UDP protocol error
     */
    @Throws(IOException::class, MqttProtocolException::class)
    protected abstract fun onStop()



    // ------------------------------------------------------------
    // Thread
    // ------------------------------------------------------------
    @Volatile
    private var run = false

    fun isRunning(): Boolean {
        return run
    }


    /**
     * Request to start reception loop thread.
     */
    fun requestStart() {
        if (isRunning()) return
        start()
    }

    /**
     * Request to stop reception loop thread.
     */
    fun requestStop() {
        run = false
    }

    /**
     * Worker: start loop thread.
     */
    protected fun start() {
        val target = makeLoopRunnable()
        val t = Thread(target, threadName)
        t.start()
    }


    @Throws(IOException::class, MqttProtocolException::class)
    private fun loop() {
        onStart()
        run = true
        while (run) {
            step()
        }
        onStop()
    }


    private fun makeLoopRunnable(): Runnable {
        return Runnable {
            try {
                loop()
            } catch (e: IOException) {
                GlobalErrorHandler.handleError(ErrorType.IO, e)
            } catch (e: MqttProtocolException) {
                GlobalErrorHandler.handleError(ErrorType.Protocol, e)
            }
        }
    }




}